### PR TITLE
New rules! (L020 - L028)

### LINKED ISSUE #9: Rule: Warn if a query uses both DISTINCT and GROUP BY
Since these two features of SQL behave very similarly, to my knowledge there is no reason to use both in the same query (I have not been able to 100% confirm this.). Even if there are occasional reasons, it's still pretty unusual and worth flagging.

If someone does this, they may be misunderstanding the behavior. The tool should suggest they eliminate `DISTINCT` from the query and do what they need using `GROUP BY`.

### LINKED ISSUE #69: Warn on missing whitespace in CTE declaration
I'd like to see a warning if there is missing whitespace after the `AS` in a CTE declaration, e.g.:
```
WITH
    count_audience_size AS( # <--- No space between AS and (
        SELECT
            user_id
        FROM
            table
    )

SELECT * FROM count_audience_size
```

### LINKED ISSUE #84: Enhancement: Raise issue if table referred to in select is not present.
Quite frequently people write queries that fail because they reference things wrong. sqlfluff won't know what columns are available (probably), but it does know what tables are available.

A query like the one below should fail because the select elements refer to tables `a` and `b` which aren't in the `FROM` statement.
```
SELECT
    a.blah,
    b.foo
FROM bar
```

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This fixes #9, fixes #69 and fixes #84, three long outstanding new rules and adds the framework for some more advanced rules in the future. All require the context of the parsed syntax to understand whether referred to tables are in the query and also whether `distinct` and `group by` are used in the same statement.

@barrywhart - both of these go quite deep into the parser and I imagine could have potential side effects - would you be able to test this PR on one of your projects and see if you agree with the results?

### CODE REVIEW

**Comment by barrywhart** [comment]
Sure, I can try out the PR. Are there particular things I should pay attention to? E.g.
* Testing it for correctness with some cases like the tickets you're addressing
* Using it to lint a large amount of SQL
* Other?

**Comment by alanmcruickshank** [comment]
@barrywhart - In order of importance:
1. If you run them on some larger codebases so you get an exceptions.
2. For rule L020, I worry that it's going to be a bit harsh, for example, also being case sensitive at the moment. Would love your feedback on if it needs to be made a bit more forgiving.
3. Given several of the rules were your suggestions, do they do what you want them to 😄

**Comment by alanmcruickshank** [comment]
I've tested these on some of our larger code bases and they look good. It's brought up a couple of other issues that I'll post separately, but unless anyone else finds any bugs I'm going to merge this in the next day or so.

**Comment by alanmcruickshank** [comment]
I realise this functionality also enables us to fix #202 which has been open a while too. I've added L024 to address this and also made the L023 more robust and generic to support that.

**Comment by barrywhart** [comment]
Nice! I will try and look at this next week. This week I got caught up trying out a constraint-checking package called [Great Expectations](https://docs.greatexpectations.io/en/latest/index.html). It supports SQL data warehouses plus Spark and Pandas. If you haven't looked at it, I recommend it.

**Comment by barrywhart** [comment]
I found a crash bug. Here's the stack trace. Details incoming...

```
    load_entry_point('sqlfluff', 'console_scripts', 'sqlfluff')()
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/bhart/.pyenv/versions/3.6.4/envs/sqlfluff-3.6.4/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/cli/commands.py", line 156, in lint
    result = lnt.lint_paths(paths, verbosity=verbose)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 762, in lint_paths
    result.add(self.lint_path(path, verbosity=verbosity, fix=fix))
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 748, in lint_path
    fix=fix, config=config))
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/linter.py", line 683, in lint_string
    lerrs, _, _, _ = crawler.crawl(parsed)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 232, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 232, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 232, in crawl
    raw_stack=raw_stack, fix=fix, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/base.py", line 186, in crawl
    raw_stack=raw_stack, memory=memory)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/rules/std.py", line 1743, in _eval
    aliases = fc.get_eventual_aliases()
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/dialects/dialect_ansi.py", line 727, in get_eventual_aliases
    buff.append(jc.get_eventual_alias())
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/dialects/dialect_ansi.py", line 672, in get_eventual_alias
    return table_expression.get_eventual_alias()
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/dialects/dialect_ansi.py", line 529, in get_eventual_alias
    return ref.reference_elements()[-1]
IndexError: list index out of range
```

**Comment by barrywhart** [comment]
This query (using dialect `bigquery`) caused the crash above.

```
WITH
audience_counts AS (
    SELECT DISTINCT
        user_id,
        list_id,
        1 AS audience
    FROM
        `project.dataset.lists_emails` AS list_emails
)

SELECT
    user_id
FROM
    audience_counts
JOIN
    `project.dataset.gdpr_safe_users`
    USING
        (user_id)
```

**Comment by barrywhart** [comment]
It seems like a pretty common issue -- the first 3 files I processed all failed with the same issue. I'll set this aside for now but will be happy to do more testing once you've looked into this issue. Thanks!

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -94,16 +94,16 @@
...
     DatatypeIdentifierSegment=ReSegment.make(r"[A-Z][A-Z0-9_]*", name='data_type_identifier', type='data_type_identifier'),
     # Maybe date parts should be more restrictive
     DatepartSegment=ReSegment.make(r"[A-Z][A-Z0-9_]*", name='date_part', type='date_part'),
-    QuotedIdentifierSegment=NamedSegment.make('double_quote', name='identifier', type='quoted_identifier'),
-    QuotedLiteralSegment=NamedSegment.make('single_quote', name='literal', type='quoted_literal'),
-    NumericLiteralSegment=NamedSegment.make('numeric_literal', name='literal', type='numeric_literal'),
+    QuotedIdentifierSegment=NamedSegment.make('double_quote', name='quoted_identifier', type='identifier'),
+    QuotedLiteralSegment=NamedSegment.make('single_quote', name='quoted_literal', type='literal'),
+    NumericLiteralSegment=NamedSegment.make('numeric_literal', name='numeric_literal', type='literal'),
     TrueSegment=KeywordSegment.make('true', name='true', type='boolean_literal'),
```
Comment (Katzmann1983) [inline src/sqlfluff/dialects/dialect_ansi.py:107]: Given that the the QuotedIdentifierSegment and NumericLiteralSegment have changed the name and type, should the TrueSegment and FalseSegment also change?

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 2]
```diff
@@ -650,7 +698,8 @@ class FromClauseSegment(BaseSegment):
                 Ref('CrossKeywordSegment'),
```
Comment (Katzmann1983) [inline src/sqlfluff/dialects/dialect_ansi.py:698]: Can you replace all lines 698 until 702 with the new JoinClauseSegment?

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_ansi.py:107]: Yes - very good point

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 2 above)
Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_ansi.py:698]: potentially yes. There *might* be a performance impact of that but I'll investigate...

**Comment by alanmcruickshank** [comment]
> It seems like a pretty common issue -- the first 3 files I processed all failed with the same issue. I'll set this aside for now but will be happy to do more testing once you've looked into this issue. Thanks!

Looks like this is a bigquery specific issue with the dots being inside the quotes in the table ref. Great catch!

**Comment by barrywhart** [comment]
Oh yeah, it's a hierarchical thing: project + dataset + table.

File: src/sqlfluff/dialects/dialect_ansi.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/dialects/dialect_ansi.py]: I've addressed this with some of the other commits below.

**Comment by alanmcruickshank** [comment]
While I was in the code, I've also added in the extra features mentioned in #277 . I think it makes much more sense to test them as a cohesive whole.

**Comment by barrywhart** [comment]
The crashing issue is fixed! I'm starting to test this PR on some of our projects.

This query is generating some warnings that raise some questions for me:
```
    SELECT
        user_id,
        list_id,
        COUNT(1) AS audience_size
    FROM
        `p.d.t1` AS lists_emails
    JOIN
        `p.d.t2`
        USING
            (user_id, list_id)
    WHERE
        active != 'D'
    GROUP BY
        user_id,
        list_id
```

Output:
```
L:   5 | P:   9 | L020 | Unqualified reference 'user_id' found in select with more than one referenced table/view.
L:   6 | P:   9 | L020 | Unqualified reference 'list_id' found in select with more than one referenced table/view.
L:   9 | P:  21 | L020 | Alias 'lists_emails' is never used in SELECT statement.
L:  12 | P:  10 | L003 | Indentation not consistent with line #11
L:  12 | P:  10 | L024 | Single whitespace expected after USING in JOIN clause.
L:  15 | P:   9 | L020 | Unqualified reference 'active' found in select with more than one referenced table/view.
L:  17 | P:   9 | L020 | Unqualified reference 'user_id' found in select with more than one referenced table/view.
L:  18 | P:   9 | L020 | Unqualified reference 'list_id' found in select with more than one referenced table/view.
```

The  "unqualified reference" warnings on lines 5 and 6 are technically not a concern -- since the two tables are `JOIN`-ed on those two fields, the fields have the same value regardless of the table.

For our data science team, I think this is a rule we may choose to disable. IIUC, the data warehouse (BigQuery or whatever) will report an error if an unqualified column reference is truly ambiguous. Our data science team works pretty independently of the engineers, and we as engineers have to be cautious about being sure new practices are worth the additional effort. I'm not saying the rule is not good, just sharing some context from our team and culture.

**Comment by barrywhart** [comment]
Performance seems much better than in the past!

I will do some specific testing of the new rules -- early next week, I hope.

**Comment by alanmcruickshank** [comment]
> Is it okay to split the USING across lines this way?

Great question, wondered what you might think. At the moment it requires *precisely* a single whitespace, not more and not less. I think you're probably right that it's a little strict. I'll expand it to make sure that it's either a single whitespace *or* that there's a newline between them.

> the fields have the same value regardless of the table.

Good point. I think in general, if there is more than one table, then good practice would be to default to always being specific. Joining with a `USING` might be a good exception however - which I think I'll add in, because in that case they literally will become the same column and it would seem bad practice to refer to them in any other way.

On the subject of `L020` and disabling it, can you have a look at some of the other cases that it catches? If you would disable *all* of them, that's easy, but if you would still want other things that it picks up to be caught, then I think I should consider breaking apart the rule into a few smaller rules which each catch something more specific.

Would appreciate any other feedback that you have!

**Comment by barrywhart** [comment]
What's the best way to understand everything `L020` does? Reading the code, running it on our SQL, or something else?

Philosophical thought... I think running a new rule on an existing project is a more stressful experience than starting with it on a new project. Seeing a zillion warnings can be overwhelming. I think sometimes I forget about the "fix" command and just run "lint" instead. Maybe having the tool do the work will change my perspective.

**Comment by alanmcruickshank** [comment]
> What's the best way to understand everything L020 does? Reading the code, running it on our SQL, or something else?

The best way is probably to read the code, but if people have to resort to that, it probably answers my question and means I should split out the bits. 👍 

> Maybe having the tool do the work will change my perspective.

I love this thought. If it does, then I would love to bring other people on that journey too. Have a think about what would do that.

**Comment by barrywhart** [comment]
I wonder if it's hard to integrate with an editor, even in a basic way... "Push a button, clean up your SQL" sounds really appealing. FWIW, we mostly use PyCharm, but I think there's some use of VSCode as well.

**Comment by alanmcruickshank** [comment]
> I wonder if it's hard to integrate with an editor, even in a basic way

I would love to do that, if you hear how, then definitely post any suggestions.

...anyway. I've split out L020 into a few sub-rules:
- L020: Table aliases should be unique within a select clause _(hopefully uncontroversial)_
- L025: Tables should not be aliased if the alias is not used in the query _(also hopefully uncontroversial)_.
- L026: No references to objects not present in the `FROM` clause _(hopefully uncontroversial, but I could imagine having annoying edge cases. Should test.)_
- L027: References should be qualified if select has more than one referenced table/view. *This* is the one that you said you might previously turn off. I've added some extra logic to exclude any columns which are referred to in a `USING` clause, because I agree that they should not need to be qualified. @barrywhart - I'd love your feedback on this rule to see if it's now more useful, or still would be on the _"turn off"_ list.
- L028: References should be consistent within statements that reference a single table. This one is one that we haven't talked about, and works well enough for now, but has the potential for strange behaviors with table functions. @barrywhart I'm curious for feedback on how it behaves with your codebase.

**Comment by barrywhart** [comment]
I'll try and evaluate these early next week. About L027, my first instinct is to try "sqlfluff fix" and see if it works well. My anxiety about enabling it is mostly cultural, because I'm an engineer and the data scientists are considered a different team. It's a new structure and I don't want to trigger any negative stereotypes about engineers caring too much about unimportant details and slowing down the ML work, if that makes sense. If it were just me or just engineers, I'd feel less nervous.

I'm definitely +1 on your decision to split out the individual rules -- your reasoning above makes sense to me. I'm looking forward to testing this again and taking a closer look, especially at L028.

**Comment by alanmcruickshank** [comment]
> It's a new structure and I don't want to trigger any negative stereotypes about engineers caring too much about unimportant details and slowing down the ML work, if that makes sense. If it were just me or just engineers, I'd feel less nervous.

Totally understand! 😃 

Full transparency: We're not using sqlfluff at all in production at tails.com yet. I'm using it to clean up a few projects, and I'm almost ready to put it in the CI pipeline for *one part* of *one project*. I think there's probably a place for a guide about how to roll out sqlfluff on a project without pissing people off. I'll make an issue for it 👍

**Comment by barrywhart** [comment]
Above, we discussed the possibility of integrating SQLFluff with an IDE. I found some technical info on how to do this with PyCharm. I'll create a separate issue for discussing this.

**Comment by alanmcruickshank** [comment]
@barrywhart - I'm going to merge this on Friday and release a new version (it's been waaaay too long since we did that). Any last feedback before I release?

**Comment by barrywhart** [comment]
That sounds good. I keep wanting to do a closer review, more testing, etc., but other work has been pulling me in different directions.
